### DISCUSSION TITLE
When using joined table inheritance, is it possible to update the parent (with its relationships), before updating children?

### DISCUSSION THREAD

**Discussion by souperk** [sqlalchemy/sqlalchemy#11064] [Usage Questions] [answered]
I have the following models:

```python
class Account(DeclarativeBase):
    __tablename__ = "account"
    __mapper_args__ = {
        "polymorphic_identity": "account",
        "polymorphic_on": "type",
    }

    id: Mapped[int] = mapped_column("id", Integer, primary_key=True)
    type: Mapped[str] = mapped_column("type", String(255), nullable=False)
    emails: Mapped[list["AccountEmail"]] = relationship(
        "AccountEmail",
        back_populates="account",
    )


class AccountEmail(DeclarativeBase):
    __tablename__ = "account_email"

    id: Mapped[int] = mapped_column("id", Integer, primary_key=True)
    account_id: Mapped[uuid.UUID] = mapped_column(
        "account_id", 
        ForeignKey("account.id"), 
        nullable=False,
    )
    account: Mapped[Account] = relationship(
        Account,
        foreign_keys=[account_id],
        back_populates="emails",
    )


class UserAccount(Account):
    __tablename__ = "user_account"
    __mapper_args__ = {
        "polymorphic_identity": "user",
    }

    account_id: Mapped[uuid.UUID] = mapped_column(
        "account_id", 
        ForeignKey("account.id"), 
        primary_key=True,
    )
    primary_email_id: Mapped[uuid.UUID] = mapped_column(
        "primary_email_id",
        ForeignKey("account_email.id"),
        nullable=False,
    )
    primary_email: Mapped[AccountEmail] = relationship(
        AccountEmail,
        foreign_keys=[primary_email_id],
        cascade="all, delete",
    )
```

When doing the following:
```python
user = UserAccount()
user.primary_email = AccountEmail()
session.add(user)
session.commit()
```

I expect the following SQL statements to be executed:
```
insert into account
insert into account_email
insert into user_account
```

Instead, I am getting the following error:
```
sqlalchemy.exc.CircularDependencyError: Circular dependency detected. (ProcessState(ManyToOneDP(UserAccount.primary_email), <UserAccount at 0x7fe4ababe4e0>, delete=False), ProcessState(ManyToOneDP(AccountEmail.account), <AccountEmail at 0x7fe4ababe7b0>, delete=False), ProcessState(OneToManyDP(Account.emails), <UserAccount at 0x7fe4ababe4e0>, delete=False), SaveUpdateState(<AccountEmail at 0x7fe4ababe7b0>), SaveUpdateState(<UserAccount at 0x7fe4ababe4e0>))
```

Also, using a debugger, I find the following dependencies:
```json
{
    "SaveUpdateState(<UserAccount at 0x7fe6ab3cf0b0>)": [
        "ProcessState(ManyToOneDP(UserAccount.primary_email), <UserAccount at 0x7fe6ab3cf0b0>, delete=False)"
    ],
    "ProcessState(OneToManyDP(Account.emails), <UserAccount at 0x7fe6ab3cf0b0>, delete=False)": [
        "SaveUpdateState(<UserAccount at 0x7fe6ab3cf0b0>)"
    ],
    "SaveUpdateState(<AccountEmail at 0x7fe6ab3ce5d0>)": [
        "ProcessState(OneToManyDP(Account.emails), <UserAccount at 0x7fe6ab3cf0b0>, delete=False)",
        "SaveUpdateState(<UserAccount at 0x7fe6ab3cf0b0>)",
        "ProcessState(ManyToOneDP(AccountEmail.account), <AccountEmail at 0x7fe6ab3ce5d0>, delete=False)"
    ],
    "ProcessState(ManyToOneDP(AccountEmail.account), <AccountEmail at 0x7fe6ab3ce5d0>, delete=False)": [
        "SaveUpdateState(<UserAccount at 0x7fe6ab3cf0b0>)"
    ],
    "ProcessState(ManyToOneDP(UserAccount.primary_email), <UserAccount at 0x7fe6ab3cf0b0>, delete=False)": [
        "SaveUpdateState(<AccountEmail at 0x7fe6ab3ce5d0>)"
    ]
}
```

The cycle seems to be the following:
1. SaveUpdateState(UserAccount)
2. ProcessState(UserAccount.primary_email)
3. SaveUpdateState(AccountEmail)
4. SaveUpdateState(UserAccount)

It seems to me that since user is tracked as a single state entry (makes sense as it's a single entity instance), I cannot achieve what I am trying to...

Also, I have tried using `post_update=True` but that would require to drop the `NOT NULL` constraint on the `primary_email_id` which I would prefer to avoid.

**Comment by zzzeek** [comment, marked as answer]
that's unfortunately not how SQLAlchemy's unit of work works, it will always handle the account/user_account rows as a single entity and done in one step, it does not have the ability to interleave related dependent rows in between the parent/child rows of a single entity.  you can either use insert() constructs directly or you'd have to make one of the two foreign key columns nullable.

**Comment by CaselIT** [comment]
if you use postgresql (or any db that supports it) you can replace the `not null` with a check constraint not null marking it 'deferred initially deferred' so that PG will allow the null value for the duration of the transaction (meaning you get a constraint violation if there is one when you try to commit, not as soon as you insert the row)

This will allow the `post_update=True` to work correctly

See https://docs.sqlalchemy.org/en/20/core/constraints.html#sqlalchemy.schema.CheckConstraint.params.deferrable and https://docs.sqlalchemy.org/en/20/core/constraints.html#sqlalchemy.schema.CheckConstraint.params.initially and https://www.postgresql.org/docs/16/sql-createtable.html#SQL-CREATETABLE-PARMS-DEFERRABLE

**Reply by souperk** [reply]
Nice idea, unfortunately postgresql does not defer check constraints. Thanks for the reference, this can be handy :pray: 

According to https://www.postgresql.org/docs/16/sql-createtable.html#SQL-CREATETABLE-PARMS-DEFERRABLE

> Currently, only UNIQUE, PRIMARY KEY, EXCLUDE, and REFERENCES (foreign key) constraints accept this clause. NOT NULL and CHECK constraints are not deferrable.

**Reply by CaselIT** [reply]
oh I was sure it did. sorry for suggesting it
